### ISSUE TITLE
[feature request] Store accumulated gradients in separate GPU or CPU memory

### ISSUE THREAD

**Issue by neonrights** [pytorch/pytorch#16700]
When using the DataParallel, accumulated gradients seem to be gathered and stored on the primary GPU.  Since the primary GPU is also used to compute forward passes and store optimizer states, it uses disproportionately more memory than other GPUs.  Due to the large size of my model, I am unable to store accumulated gradients in a single GPU while using the same GPU to compute future losses and gradients.

I am wondering if it is possible to gather gradients computed in a DataParallel model to a separate, dedicated GPU, or to even store them in CPU memory along with the optimizer.

More details can be found on this [pytorch topic](https://discuss.pytorch.org/t/uneven-gpu-utilization-during-training-backpropagation/36117).

**Comment by mrshenli** [comment]
If you prefer to stick with `DataParallel` instead of using `DistributedDataParallel`, have you tried the method mentioned in [this post](https://medium.com/huggingface/training-larger-batches-practical-tips-on-1-gpu-multi-gpu-distributed-setups-ec88c3e51255)? @rasbt also suggested something similar in the forum.

**Comment by rasbt** [comment]
I think the gradient accumulation approach might work as a workaround together with DataParallel. Also DistributedParrallel might be a solution but comes with much more overhead. 

However, while these are workarounds, I am still wondering if a modification to DataParallel can be considered to help with this issue more naturally. E.g., assume your machine has *k* devices

- define input device ID's 0 to (*k*-1) as input devices via the existing input `device_ids` param
- define the *k*th device as output device via the existing `output_device` param (by default, it's currently `default: device_ids[0]` and one can only choose device from the `device_ids` list.

So, the only change would basically be selecting a GPU as output device that is not used for training.

**Comment by neonrights** [comment]
My particular issue is that gradients are gathered and stored in ```device_ids[0]```.  This results in ```device_ids[0]``` using disproportionately more memory since it also computes a mini-batch with other devices. This is exacerbated when performing and update step with an optimizer such as ```Adam``` since it also uses memory on ```device_ids[0]```.  This leads to memory bottle-necking despite plenty more memory being available on other devices.

I have tried specifying a separate device as output as @rasbt suggested, however this seems only gather the loss computed during my forward pass to ```output_device```.  When calling ```backward()``` on the outputted loss, gradients still seem to be gathered and stored in ```device_ids[0]```.

I have also tried using the ```encoding.parallel``` module in the post linked.  This seems to only parallelize the loss calculation.  Since I include my loss calculation in the forward pass, ```DataParallel``` seems to already do this.  I hesitate to try other methods such as checkpointing since it would make my training time infeasibly slow.

I'll see if ```DistributedDataParallel``` is a solution once I have it implemented.

**Comment by rasbt** [comment]
Yes, that's basically what I was proposing above. You may not even need a new variable but can do that on the GPU that is specified as via `output_device`. The reason why this currently doesn't work is that the `output_device` must be in the device id list.

**Comment by neonrights** [comment]
From my understanding, ```DataParallel``` executes a forward and backward pass like so, where the replicated model is initially on ```device_ids[0]```.

![DataParallel execution](https://cdn-images-1.medium.com/max/1000/1*FpDHkWJhkLL7KxU01Lf9Lw.png)

Could the proposed behavior be accomplised by altering the second step and replicating the model from the ```output_device```?  Since the extra copy on the ```output_device``` would not have received a mini-batch during the scatter operation, it can be ignored during parallel application of the forward and backwards pass.  Once gradients are calculated, they should be gathered back into the ```output_device```.  However, the model will now be required to be on ```output_device``` while each mini-batch would have to be on ```device_ids[0]```.

**Comment by soumith** [comment]
if you use DistributedDataParallel, you dont have separate storage of anything. The gradients are simply all-reduced across all devices with no extra memory (except a very very small scratch memory).

@rasbt why did you think DDP is more overhead? It's faster, more memory efficient and better in all possible ways.

**Comment by rasbt** [comment]
Sorry, that was unclear. From a computational perspective, I guess there are several advantages of DistributedDataParallel over DataParallel. I was only referring to "code implementation" overhead on the user-side, i.e., that where DataParallel only required changing 1-2 lines of code, DistributedDataParallel required more substantial changes, which was in my personal use cases too much hassle since I usually only use a single machine (either a 4 GPU workstation or 8 GPU server with the dataset on its local SSD), and when I understood correctly, DistributedDataParallel was more geared towards using multiple machines -- although, browsing through the docs, it seems that it can be used for a single machine as well.

So, when I refer to overhead, I mean that it's extremely convenient for my use cases to use the regular DataParallel, as I can simply nest it in simple if-else clauses and then toggle whether I want to run it on the available GPUs or just one.

Good that you mentioned it though, when I was thinking of DistributedDataParallel, I think I was mostly thinking of the API that now seems to be deprecated: https://pytorch.org/docs/stable/distributed_deprecated.html

I have yet to try the "new" API (https://pytorch.org/docs/stable/distributed.html).
